### PR TITLE
feat: Add a script that calculates the multiplicative probability error

### LINKED ISSUE #2: Test script to verify new models compatibility with RL pipeline
**Is your feature request related to a problem? Please describe.**

Different libraries/environments for training vs generation can lead to unexpected errors which are hard to debug since RL is robust enough that just observing rewards isn't a good enough check.

**Describe the solution you'd like**

We have already added a "Adding New Models" guide which explains the importance of tracking logprob errors. The guide should be accompanied by a script that the user is expected to run every time they bring in a new model with RL pipeline which essentially tests if training and inference frameworks for a new model are compatible.

### PR DESCRIPTION
**Pull request by terrykong**
- Adds script/docs to test new models
- Made a change to normalize `prepare_for_lp_inference`/`prepare_for_generation`/`finish_generation` so that it's all just called `prepare_for_inference` and `finish_inference`. This led to less conditionals checking if I had a HFPolicy or VllmGeneration
- 

---

Closes #2 

---

image

`uv run examples/stress_test_model.py` (llama 3.2 1B)
```
SUMMARY RESULTS TABLE
                                                                                                         
  ISL (avg)   OSL (avg)   Max New Tokens   Data Type   Generation Method   Batch Size   HF vs vllm(gen)  
 ─────────────────────────────────────────────────────────────────────────────────────────────────────── 
     233.00      512.00              512   numbers     greedy                       8            1.0001  
      86.25      465.19              512   random      greedy                       8            1.0034  
      10.88      399.00              512   literal     greedy                       8            1.0053  
     233.00      512.00              512   numbers     default                      8            1.0001  
      86.25      465.19              512   random      default                      8            1.0034  
      10.88      399.00              512   literal     default                      8            1.0053  
     233.00     7960.00             8192   numbers     greedy                       8            1.0000  
      86.25     6630.31             8192   random      greedy                       8            1.0025  
      10.88     6154.38             8192   literal     greedy                       8            1.0032  
     233.00     7960.00             8192   numbers     default                      8            1.0000  
      86.25     6630.31             8192   random      default                      8            1.0025  
      10.88     6154.38             8192   literal     default                      8            1.0032  
```

`uv run examples/stress_test_model.py generation.backend=hf` (llama 3.2 1B)
```
SUMMARY RESULTS TABLE
                                                                                                       
  ISL (avg)   OSL (avg)   Max New Tokens   Data Type   Generation Method   Batch Size   HF vs hf(gen)  
 ───────────────────────────────────────────────────────────────────────────────────────────────────── 
     233.00      512.00              512   numbers     greedy                       8          1.0001  
      88.12      456.50              512   random      greedy                       8          1.0045  
      10.88      371.88              512   literal     greedy                       8          1.0058  
     233.00      512.00              512   numbers     default                      8          1.0047  
      88.12      469.31              512   random      default                      8          1.0151  
      10.88      417.38              512   literal     default                      8          1.0172  
     233.00     8192.00             8192   numbers     greedy                       8          1.0000  
      88.12     6731.00             8192   random      greedy                       8          1.0037  
      10.88     5171.88             8192   literal     greedy                       8          1.0045  
     233.00     6038.56             8192   numbers     default                      8          1.0068  
      88.12     3223.12             8192   random      default                      8          1.0133  
      10.88     3381.50             8192   literal     default                      8          1.0138  
```

`uv run examples/stress_test_model.py generation.model_name=Qwen/Qwen2.5-Math-1.5B-Instruct generation.vllm_cfg.max_model_len=4096`
```
SUMMARY RESULTS TABLE
                                                                                                         
  ISL (avg)   OSL (avg)   Max New Tokens   Data Type   Generation Method   Batch Size   HF vs vllm(gen)  
 ─────────────────────────────────────────────────────────────────────────────────────────────────────── 
     386.25      512.00              512   numbers     greedy                       8            1.0001  
      88.50      512.00              512   random      greedy                       8            1.0007  
      10.88      419.50              512   literal     greedy                       8            1.0072  
     386.25      512.00              512   numbers     default                      8            1.0001  
      88.50      512.00              512   random      default                      8            1.0007  
      10.88      419.50              512   literal     default                      8            1.0072  
     386.25     3710.75             4096   numbers     greedy                       8            1.0001  
      88.50     4008.50             4096   random      greedy                       8            1.0166  
      10.88     2656.00             4096   literal     greedy                       8            1.0043  
     386.25     3710.75             4096   numbers     default                      8            1.0001  
      88.50     4008.50             4096   random      default                      8            1.0166  
      10.88     2656.00             4096   literal     default                      8            1.0043 
```

Checkpointing will change in #..., but can use this for v0.1.1 checkpoints:
```python
import os
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM
import sys
import json

input_dir = sys.argv[1]
output_dir = sys.argv[2]

with open(os.path.join(input_dir, "config.json"), "r") as f:
    config = json.load(f)

reference_model = config["policy"]["model_name"]

weights_path = os.path.join(input_dir, "policy.pt")

tokenizer = AutoTokenizer.from_pretrained(reference_model)
model = AutoModelForCausalLM.from_pretrained(reference_model)

state_dict = torch.load(weights_path)
model.load_state_dict(state_dict)

model.save_pretrained(output_dir)
tokenizer.save_pretrained(output_dir)
```

`uv run examples/stress_test_model.py generation.model_name=/path/to/model/converted/from/script/above`

### CODE REVIEW

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 1]
```diff
@@ -706,7 +720,11 @@ def report_device_id(self) -> str:
     def get_weight_ipc_handles(self, offload_model=True):
         from torch.multiprocessing.reductions import reduce_tensor
 
+        # Move the model to the GPU just in case it was previously offloaded
+        self.model.to("cuda")
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/hf_policy.py:755]: In the stress script, since I'm not training, the model was left on the CPU when I tried to refit and led to an error. Seems safer to always move it to GPU before trying to gather and pass ipc handle

File: nemo_reinforcer/distributed/batched_data_dict.py [hunk 2]
```diff
@@ -153,6 +153,33 @@ def shard_by_batch_size(
...
+        >>> len(sharded)
+        2
+        >>> assert sharded[0]["list_data"] == ["A", "B", "C", "D"]
+        >>> assert sharded[1]["list_data"] == ["E", "F", "G", "H"]
+        >>>
+        >>> # These are equivalent
+        >>> assert sharded == batch.shard_by_batch_size(shards=2, batch_size=8)
+        >>> assert sharded == batch.shard_by_batch_size(shards=2, batch_size=None)
+        >>>
+        >>> sharded = batch.shard_by_batch_size(shards=2, batch_size=4)
```
Comment (terrykong) [inline nemo_reinforcer/distributed/batched_data_dict.py:177]: added some doctests, since this last behavior wasn't obvious to me

File: nemo_reinforcer/distributed/batched_data_dict.py [hunk 3]
```diff
@@ -253,15 +280,15 @@ def repeat_interleave(self, num_repeats: int) -> "BatchedDataDict":
         return repeated_batch
 
     def make_microbatch_iterator(
-        self, microbatch_size: int
+        self, max_microbatch_size: int
```
Comment (terrykong) [inline nemo_reinforcer/distributed/batched_data_dict.py:283]: changes this to max_microbatch_size since we should allow making an iterator when the batch size is not divisible. this helps so we don't have to recreate the hfpolicy when calculating logprobs when we alternate between different inference batch sizes

File: nemo_reinforcer/models/generation/vllm.py [hunk 4]
```diff
@@ -195,13 +196,17 @@ def is_alive(self):
         return True
 
     def generate(
-        self, data: BatchedDataDict[GenerationDatumSpec], greedy: bool = False
+        self,
+        data: BatchedDataDict[GenerationDatumSpec],
+        greedy: bool = False,
+        sampling_param_overrides: Optional[dict] = None,
```
Comment (terrykong) [inline nemo_reinforcer/models/generation/vllm.py:202]: passes in sampling param overrides so we don't need to recreate vllmgeneration just to try a different set of temperature/top_p/topk

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 5]
```diff
@@ -141,7 +141,7 @@ def do_fsdp(model):
         else:
             self.optimizer = None
 
-        if "scheduler" in self.cfg:
+        if self.optimizer and "scheduler" in self.cfg:
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/hf_policy.py]: some fixes

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 6]
```diff
@@ -470,12 +470,19 @@ def get_reference_policy_logprobs(self, data: BatchedDataDict) -> BatchedDataDic
         return return_data
 
     def generate(
-        self, data: BatchedDataDict[GenerationDatumSpec], greedy: bool = False
+        self,
+        data: BatchedDataDict[GenerationDatumSpec],
+        greedy: bool = False,
+        generation_batch_size: int | None = None,
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/hf_policy.py:492]: would have liked to not have to add this arg here and have it's behavior be similar to vllm's generation. think i should just remove and infer it like for vllmgeneration @SahilJain314 ?

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 7]
```diff
@@ -729,10 +747,18 @@ def get_weight_ipc_handles(self, offload_model=True):
...
+
+        Args:
+            offload_optimizer_and_buffers: If True, offload the optimizer and buffers to the CPU.
+              True saves more memory and should be used if you only need to calculate logprobs (no generation).
+              Set to False if you need to generate since model buffers should remain on the GPU.
+        """
         self.model.to("cuda")
         self.model.eval()
-        self.offload_before_refit()
+        if offload_optimizer_and_buffers:
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/hf_policy.py:791]: i had to introduce this b/c for log prob inference, i can offload the buffers, but for generation, i couldn't do that because some buffers were necessary for generation (i think it was was inverse freq related to rope)
